src/fromager/sources.py

```diff
@@ -119,6 +140,136 @@ def resolve_source(
     return str(url), version
 
 
+def resolve_version_from_git_url(
+    ctx: context.WorkContext,
+    req: Requirement,
+) -> tuple[str, Version]:
+    "Return path to the cloned git repository and the package version."
+
+    if not req.url.startswith("git+"):
+        raise ValueError(f"unable to handle URL scheme in {req.url} from {req}")
+
+    # We start by not knowing where we would put the source because we don't
+    # know the version.
+    working_src_dir = ""
+    version: Version | None = None
+
+    # Clean up the URL so we can parse it
+    reduced_url = req.url[len("git+") :]
+    parsed_url = urlparse(reduced_url)
+
+    # Save the URL that we think we will use for cloning. This might change
+    # later if the path has a tag or branch in it.
+    url_to_clone = reduced_url
+    need_to_clone = False
+
+    # If the URL includes an @ with text after it, we use that as the reference
+    # to clone, but by default we take the default branch.
+    git_ref: str | None = None
+
+    if "@" not in parsed_url.path:
+        # If we have no reference, we know we are going to have to clone the
+        # repository to figure out the version to use.
+        need_to_clone = True
+    else:
+        # If we have a reference, it might be a valid python version string, or
+        # not. It _must_ be a valid git reference. If it can be parsed as a
+        # valid python version, we assume the tag points to source that will
+        # think that is its version, so we allow reusing an existing cloned repo
+        # if there is one.
+        new_path, _, git_ref = parsed_url.path.rpartition("@")
+        url_to_clone = parsed_url._replace(path=new_path).geturl()
+        try:
+            version = Version(git_ref)
+        except ValueError:
+            logger.info(
+                "%s: could not parse %r as a version, cloning to get the version",
+                req.name,
+                git_ref,
+            )
+            need_to_clone = True
+        else:
+            logger.info("%s: URL %s includes version %s", req.name, req.url, version)
+            working_src_dir = (
+                ctx.work_dir / f"{req.name}-{version}" / f"{req.name}-{version}"
+            )
+            if not working_src_dir.exists():
+                need_to_clone = True
+            else:
+                if ctx.cleanup:
+                    logger.debug("%s: cleaning up %s", req.name, working_src_dir)
+                    shutil.rmtree(working_src_dir)
+                    need_to_clone = True
+                else:
+                    logger.info("%s: reusing %s", req.name, working_src_dir)
+
+    if need_to_clone:
+        with tempfile.TemporaryDirectory() as tmpdir:
+            clone_dir = pathlib.Path(tmpdir) / "src"
+            gitutils.git_clone(
+                ctx=ctx,
+                req=req,
+                output_dir=clone_dir,
+                repo_url=url_to_clone,
+                ref=git_ref,
```

Comment:
`pip install git+http` always pulls all sub modules. Maybe we should do the same? Or at least document that we are not pulling submodules and add a helper function to update them?

pip's git handling code is in https://github.com/pypa/pip/blob/main/src/pip/_internal/vcs/git.py

Reply:
I could come back and add git submodule handling in another PR. We don't need it for what we're building today.

Reply:
#563